design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,188 @@
+* The non-persistent nature of UUIDs may trigger redundant license activation events during VM lifetime i.e. Windows key activation re-occur when power off and power on the VM.
+
+## Goals
+* Improve the uniqueness of the VMI firmware UUID to become independent of the VM name and namespace.
+* Maintain the UUID persistence over the VM lifecycle.
+* Maintain backward compatibility
+* Maintain UUID persistence with VM backup/restore.
+
+## Non Goals
+
+
+## Definition of Users
+End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.
+
+## User Stories
+As an end-user, I expect my VMI to maintain its identity across restarts.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+
+This section provides a range of alternative solutions for implementing a persistent,
+universally unique firmware UUID for Virtual Machines Instances (VMIs) in KubeVirt.
+
+### 1. Persist UUID in the VM's Spec Field
+
+**Description:** The UUID is generated when the VMI is being created and is then saved to the VM's spec field.
+
+**Pros:**
+- Straightforward implementation.
+- Avoids the need to introduce new API fields.
+- Fully backward compatible.
+
+**Cons:**
+- Potential misuse of the spec field, which ideally should only reflect the user’s intended configuration.
+- The majority of users may not consider the UUID part of the desired VM state, making this field irrelevant from their perspective.
+- Can lead to longer and more complex VM definitions, impacting readability and management.
+
+---
+
+### 2. Generate a Default UUID via Webhook and Set it to the Spec Field
+
+**Description:** A webhook generates a UUID if the user does not provide one. This UUID is then set in the VM's spec field.
+
+**Pros:**
+- Retains backward compatibility and avoids introducing new API fields.
+
+**Cons:**
+- Same concerns as above about the spec field.
+- Using webhooks to assign defaults can degrade performance, especially with high volumes of VMs, as `virt-api` may become a bottleneck.
+- May compromise scalability, as generating a UUID via webhook adds overhead when multiple VMs start simultaneously.
+
+---
+
+### 3. Create a New Firmware UUID Field in VM Status
+
+**Description:** A new field is added under VM status to store the generated firmware UUID upon first VM boot.
+
+**Pros:**
+- Keeps the spec clean, avoiding unnecessary fields in the user-defined configuration.
+- Provides a clear separation between user configuration (spec) and system-generated information (status).
+
+**Cons:**
+- Introduces a new API field, which could increase API surface area and requires long-term support.
+- New fields may impact system performance and resource load.
+- Adds to API complexity and may reduce readability.
+
+*Note:* To minimize this con, consider setting the UUID as a condition under VM status instead of introducing a separate field.
+
```

Comment:
Again how do we ensure the existing VMs retain stable UUID?

Reply:
I think the flow of option 3 is not well explained - at least from what I had in mind.
I see this working the following: 

VM controller: 
   On start:
       if the user explicitly set the `vm.spec.template.spec.Firmware.UUID` - propagate this to `vmi.spec.Firmware.UUID`
       Otherwise (user didn't set firmware UUID explicitly):
            sets the `vmi.spec.Firmware.UUID` to `vm.metadata.uid` if `vm.Status.FirmwareUUID` is empty (because it's a newly created VM - no VMI existed before)
            if vm.status.FirmwareUUID is set - propagate it to `vmi.spec.Firmware.UUID`

    On Sync:
        if `vm.status.FirmwareUUID` is empty set it using `vmi.spec.Firmware.UUID` 

This approach preserves backward compatibility and is still gitops friendly

Reply:
@xpivarc @vladikr I think we can re-use vTPM backend storage approach here, instead of using the VM status as the carrier for this data, WDYT?

Reply:
As I imagine it, by using the same approach as with vTPM backend storage it would be easy to manage the VM Cloning, snapshot and restore aspects, i.e. we would not clone the UUID PVC as it should be unique just like the vTPM